### PR TITLE
Migrate conv3d from TH to ATen (CPU)

### PR DESCRIPTION
**Pull request by andreaskoepf**
This is a port of the VolumetricConvolutionMM TH (CPU) implementation to ATen as `slow_conv3d`.

- [x] unfolded3d_copy & unfolded3d_acc
- [x] forward
- [x] backward
- [x] basic sanity cross check with 1.3 impl
- [x] systematic testing
- [x] performance comparison & optimization

Script used for performance testing: [benchmark_conv3d.py](https://gist.github.com/andreaskoepf/8865eea4bb05220f78fc6d9d408c49fc)

### CODE REVIEW

File: aten/src/ATen/native/ConvolutionMM3d.cpp [hunk 1]
```diff
@@ -0,0 +1,655 @@
+#include <ATen/ATen.h>
+#include <ATen/Dispatch.h>
+#include <ATen/Parallel.h>
+#include <ATen/TensorUtils.h>
+#include <ATen/core/grad_mode.h>
+#include <ATen/div_rtn.h>
+#include <ATen/native/Unfold3d.h>
+
+#define CONV3D_GRAIN_SALT 20
```
Comment (ezyang) [inline aten/src/ATen/native/ConvolutionMM3d.cpp]: nit: no reason to use a macro here, constexpr int will work just fine

**Review by ezyang — APPROVED** [review]

**Comment by ezyang** [comment]
Needs a rebase to avoid conflict with the landed conv2d.

**Comment by andreaskoepf** [comment]
## Performance Test Results

Master and PR feature branch builds were performed with:
`BLAS=MKL USE_NNPACK=0 DEBUG=0 USE_CUDA=0 BUILD_TEST=0 BUILD_CAFFE2_OPS=0 python setup.py develop`

### Forward

The percentage value in parentheses is new_duration / old_duration.
The durations are averaged over 100 runs in a loop.

```
in: 3x10x32x32; ch: 32; k: 5x7x7; pad: 4; st: 1; new: 60.621ms; old: 66.188ms  (91.59%  ) 
in: 3x10x32x32; ch: 32; k: 5x7x7; pad: 4; st: 2; new:  4.481ms; old:  5.268ms  (85.06%  ) 
in: 3x10x32x32; ch: 32; k: 5x7x7; pad: 2; st: 1; new: 34.342ms; old: 36.759ms  (93.42%  ) 
in: 3x10x32x32; ch: 32; k: 5x7x7; pad: 2; st: 2; new:  2.785ms; old:  3.288ms  (84.70%  ) 
in: 3x10x32x32; ch: 32; k: 5x7x7; pad: 1; st: 1; new: 25.437ms; old: 27.407ms  (92.81%  ) 
in: 3x10x32x32; ch: 32; k: 5x7x7; pad: 1; st: 2; new:  2.021ms; old:  2.238ms  (90.30%  ) 
in: 3x10x32x32; ch: 32; k: 4x4x5; pad: 4; st: 1; new: 26.132ms; old: 27.458ms  (95.17%  ) 
in: 3x10x32x32; ch: 32; k: 4x4x5; pad: 4; st: 2; new:  2.143ms; old:  2.424ms  (88.41%  ) 
in: 3x10x32x32; ch: 32; k: 4x4x5; pad: 2; st: 1; new: 14.904ms; old: 16.098ms  (92.58%  ) 
in: 3x10x32x32; ch: 32; k: 4x4x5; pad: 2; st: 2; new:  1.471ms; old:  1.572ms  (93.58%  ) 
in: 3x10x32x32; ch: 32; k: 4x4x5; pad: 1; st: 1; new:  6.405ms; old:  9.935ms  (64.47%  ) 
in: 3x10x32x32; ch: 32; k: 4x4x5; pad: 1; st: 2; new:  1.129ms; old:  1.115ms  (101.26% ) 
in: 3x10x32x32; ch: 32; k: 3x3x3; pad: 4; st: 1; new: 11.344ms; old:  7.909ms  (143.43% ) 
in: 3x10x32x32; ch: 32; k: 3x3x3; pad: 4; st: 2; new:  1.000ms; old:  0.978ms  (102.25% ) 
in: 3x10x32x32; ch: 32; k: 3x3x3; pad: 2; st: 1; new:  4.330ms; old:  4.951ms  (87.46%  ) 
in: 3x10x32x32; ch: 32; k: 3x3x3; pad: 2; st: 2; new:  0.684ms; old:  0.683ms  (100.15% ) 
in: 3x10x32x32; ch: 32; k: 3x3x3; pad: 1; st: 1; new:  3.568ms; old:  3.974ms  (89.78%  ) 
in: 3x10x32x32; ch: 32; k: 3x3x3; pad: 1; st: 2; new:  0.638ms; old:  0.466ms  (136.91% ) 
in: 3x10x32x32; ch: 64; k: 5x7x7; pad: 4; st: 1; new: 72.474ms; old: 76.466ms  (94.78%  ) 
in: 3x10x32x32; ch: 64; k: 5x7x7; pad: 4; st: 2; new:  6.832ms; old:  7.085ms  (96.43%  ) 
in: 3x10x32x32; ch: 64; k: 5x7x7; pad: 2; st: 1; new: 41.034ms; old: 43.965ms  (93.33%  ) 
in: 3x10x32x32; ch: 64; k: 5x7x7; pad: 2; st: 2; new:  3.755ms; old:  4.104ms  (91.50%  ) 
in: 3x10x32x32; ch: 64; k: 5x7x7; pad: 1; st: 1; new: 28.545ms; old: 32.812ms  (87.00%  ) 
in: 3x10x32x32; ch: 64; k: 5x7x7; pad: 1; st: 2; new:  2.684ms; old:  2.823ms  (95.08%  ) 
in: 3x10x32x32; ch: 64; k: 4x4x5; pad: 4; st: 1; new: 30.928ms; old: 33.706ms  (91.76%  ) 
in: 3x10x32x32; ch: 64; k: 4x4x5; pad: 4; st: 2; new:  3.076ms; old:  3.328ms  (92.43%  ) 
in: 3x10x32x32; ch: 64; k: 4x4x5; pad: 2; st: 1; new: 17.668ms; old: 19.229ms  (91.88%  ) 
in: 3x10x32x32; ch: 64; k: 4x4x5; pad: 2; st: 2; new:  2.087ms; old:  1.982ms  (105.30% ) 
in: 3x10x32x32; ch: 64; k: 4x4x5; pad: 1; st: 1; new: 12.368ms; old: 10.762ms  (114.92% ) 
in: 3x10x32x32; ch: 64; k: 4x4x5; pad: 1; st: 2; new:  1.613ms; old:  1.680ms  (96.01%  ) 
in: 3x10x32x32; ch: 64; k: 3x3x3; pad: 4; st: 1; new: 15.861ms; old: 11.434ms  (138.72% ) 
in: 3x10x32x32; ch: 64; k: 3x3x3; pad: 4; st: 2; new:  1.606ms; old:  1.419ms  (113.18% ) 
in: 3x10x32x32; ch: 64; k: 3x3x3; pad: 2; st: 1; new:  6.131ms; old:  6.846ms  (89.56%  ) 
in: 3x10x32x32; ch: 64; k: 3x3x3; pad: 2; st: 2; new:  0.996ms; old:  0.980ms  (101.63% ) 
in: 3x10x32x32; ch: 64; k: 3x3x3; pad: 1; st: 1; new:  5.246ms; old:  5.764ms  (91.01%  ) 
in: 3x10x32x32; ch: 64; k: 3x3x3; pad: 1; st: 2; new:  0.937ms; old:  0.644ms  (145.50% ) 
in: 3x10x32x32; ch: 24; k: 5x7x7; pad: 4; st: 1; new: 52.285ms; old: 63.336ms  (82.55%  ) 
in: 3x10x32x32; ch: 24; k: 5x7x7; pad: 4; st: 2; new:  4.005ms; old:  4.951ms  (80.89%  ) 
in: 3x10x32x32; ch: 24; k: 5x7x7; pad: 2; st: 1; new: 29.539ms; old: 35.971ms  (82.12%  ) 
in: 3x10x32x32; ch: 24; k: 5x7x7; pad: 2; st: 2; new:  2.357ms; old:  2.889ms  (81.59%  ) 
in: 3x10x32x32; ch: 24; k: 5x7x7; pad: 1; st: 1; new: 22.002ms; old: 26.085ms  (84.35%  ) 
in: 3x10x32x32; ch: 24; k: 5x7x7; pad: 1; st: 2; new:  1.600ms; old:  2.108ms  (75.90%  ) 
in: 3x10x32x32; ch: 24; k: 4x4x5; pad: 4; st: 1; new: 22.817ms; old: 26.266ms  (86.87%  ) 
in: 3x10x32x32; ch: 24; k: 4x4x5; pad: 4; st: 2; new:  1.729ms; old:  2.415ms  (71.59%  ) 
in: 3x10x32x32; ch: 24; k: 4x4x5; pad: 2; st: 1; new:  7.047ms; old:  9.497ms  (74.20%  ) 
in: 3x10x32x32; ch: 24; k: 4x4x5; pad: 2; st: 2; new:  1.158ms; old:  1.421ms  (81.49%  ) 
in: 3x10x32x32; ch: 24; k: 4x4x5; pad: 1; st: 1; new:  5.353ms; old:  8.151ms  (65.67%  ) 
in: 3x10x32x32; ch: 24; k: 4x4x5; pad: 1; st: 2; new:  0.973ms; old:  1.093ms  (89.02%  ) 
in: 3x10x32x32; ch: 24; k: 3x3x3; pad: 4; st: 1; new:  5.723ms; old:  7.408ms  (77.25%  ) 
in: 3x10x32x32; ch: 24; k: 3x3x3; pad: 4; st: 2; new:  0.957ms; old:  0.824ms  (116.14% ) 
in: 3x10x32x32; ch: 24; k: 3x3x3; pad: 2; st: 1; new:  3.471ms; old:  4.474ms  (77.58%  ) 
in: 3x10x32x32; ch: 24; k: 3x3x3; pad: 2; st: 2; new:  0.580ms; old:  0.595ms  (97.48%  ) 
in: 3x10x32x32; ch: 24; k: 3x3x3; pad: 1; st: 1; new:  2.816ms; old:  3.674ms  (76.65%  ) 
in: 3x10x32x32; ch: 24; k: 3x3x3; pad: 1; st: 2; new:  0.485ms; old:  0.542ms  (89.48%  )
```

### Backward

Results are not averaged, just one pass per test (single tests may be outliers).

```
in: 3x10x32x32; ch: 32; k: 5x7x7; pad: 4; st: 1; new: 141.168ms;  old: 262.168ms (53.85% )
in: 3x10x32x32; ch: 32; k: 5x7x7; pad: 4; st: 2; new: 17.636ms ;  old: 25.355ms  (69.56% )
in: 3x10x32x32; ch: 32; k: 5x7x7; pad: 2; st: 1; new: 78.929ms ;  old: 111.551ms (70.76% )
in: 3x10x32x32; ch: 32; k: 5x7x7; pad: 2; st: 2; new: 9.695ms  ;  old: 13.765ms  (70.43% )
in: 3x10x32x32; ch: 32; k: 5x7x7; pad: 1; st: 1; new: 57.773ms ;  old: 74.906ms  (77.13% )
in: 3x10x32x32; ch: 32; k: 5x7x7; pad: 1; st: 2; new: 6.466ms  ;  old: 4.435ms   (145.79%)
in: 3x10x32x32; ch: 32; k: 4x4x5; pad: 4; st: 1; new: 67.250ms ;  old: 98.821ms  (68.05% )
in: 3x10x32x32; ch: 32; k: 4x4x5; pad: 4; st: 2; new: 12.458ms ;  old: 16.133ms  (77.22% )
in: 3x10x32x32; ch: 32; k: 4x4x5; pad: 2; st: 1; new: 48.428ms ;  old: 61.945ms  (78.18% )
in: 3x10x32x32; ch: 32; k: 4x4x5; pad: 2; st: 2; new: 2.902ms  ;  old: 7.580ms   (38.28% )
in: 3x10x32x32; ch: 32; k: 4x4x5; pad: 1; st: 1; new: 32.595ms ;  old: 45.622ms  (71.45% )
in: 3x10x32x32; ch: 32; k: 4x4x5; pad: 1; st: 2; new: 2.925ms  ;  old: 5.238ms   (55.84% )
in: 3x10x32x32; ch: 32; k: 3x3x3; pad: 4; st: 1; new: 24.250ms ;  old: 57.998ms  (41.81% )
in: 3x10x32x32; ch: 32; k: 3x3x3; pad: 4; st: 2; new: 4.989ms  ;  old: 7.564ms   (65.96% )
in: 3x10x32x32; ch: 32; k: 3x3x3; pad: 2; st: 1; new: 12.193ms ;  old: 40.394ms  (30.19% )
in: 3x10x32x32; ch: 32; k: 3x3x3; pad: 2; st: 2; new: 4.210ms  ;  old: 3.260ms   (129.14%)
in: 3x10x32x32; ch: 32; k: 3x3x3; pad: 1; st: 1; new: 8.290ms  ;  old: 33.375ms  (24.84% )
in: 3x10x32x32; ch: 32; k: 3x3x3; pad: 1; st: 2; new: 3.431ms  ;  old: 3.175ms   (108.06%)
in: 3x10x32x32; ch: 64; k: 5x7x7; pad: 4; st: 1; new: 200.007ms;  old: 223.948ms (89.31% )
in: 3x10x32x32; ch: 64; k: 5x7x7; pad: 4; st: 2; new: 30.930ms ;  old: 33.613ms  (92.02% )
in: 3x10x32x32; ch: 64; k: 5x7x7; pad: 2; st: 1; new: 112.990ms;  old: 132.117ms (85.52% )
in: 3x10x32x32; ch: 64; k: 5x7x7; pad: 2; st: 2; new: 16.731ms ;  old: 16.601ms  (100.78%)
in: 3x10x32x32; ch: 64; k: 5x7x7; pad: 1; st: 1; new: 79.370ms ;  old: 95.797ms  (82.85% )
in: 3x10x32x32; ch: 64; k: 5x7x7; pad: 1; st: 2; new: 9.834ms  ;  old: 8.640ms   (113.82%)
in: 3x10x32x32; ch: 64; k: 4x4x5; pad: 4; st: 1; new: 93.819ms ;  old: 152.575ms (61.49% )
in: 3x10x32x32; ch: 64; k: 4x4x5; pad: 4; st: 2; new: 8.077ms  ;  old: 22.544ms  (35.83% )
in: 3x10x32x32; ch: 64; k: 4x4x5; pad: 2; st: 1; new: 49.569ms ;  old: 90.729ms  (54.63% )
in: 3x10x32x32; ch: 64; k: 4x4x5; pad: 2; st: 2; new: 6.423ms  ;  old: 16.405ms  (39.15% )
in: 3x10x32x32; ch: 64; k: 4x4x5; pad: 1; st: 1; new: 35.352ms ;  old: 55.862ms  (63.28% )
in: 3x10x32x32; ch: 64; k: 4x4x5; pad: 1; st: 2; new: 5.116ms  ;  old: 6.796ms   (75.28% )
in: 3x10x32x32; ch: 64; k: 3x3x3; pad: 4; st: 1; new: 33.686ms ;  old: 127.140ms (26.50% )
in: 3x10x32x32; ch: 64; k: 3x3x3; pad: 4; st: 2; new: 7.221ms  ;  old: 18.412ms  (39.22% )
in: 3x10x32x32; ch: 64; k: 3x3x3; pad: 2; st: 1; new: 28.484ms ;  old: 61.426ms  (46.37% )
in: 3x10x32x32; ch: 64; k: 3x3x3; pad: 2; st: 2; new: 6.027ms  ;  old: 8.585ms   (70.20% )
in: 3x10x32x32; ch: 64; k: 3x3x3; pad: 1; st: 1; new: 14.303ms ;  old: 41.681ms  (34.32% )
in: 3x10x32x32; ch: 64; k: 3x3x3; pad: 1; st: 2; new: 3.027ms  ;  old: 4.883ms   (61.99% )
in: 3x10x32x32; ch: 24; k: 5x7x7; pad: 4; st: 1; new: 121.008ms;  old: 162.624ms (74.41% )
in: 3x10x32x32; ch: 24; k: 5x7x7; pad: 4; st: 2; new: 9.534ms  ;  old: 10.026ms  (95.09% )
in: 3x10x32x32; ch: 24; k: 5x7x7; pad: 2; st: 1; new: 80.495ms ;  old: 83.732ms  (96.13% )
in: 3x10x32x32; ch: 24; k: 5x7x7; pad: 2; st: 2; new: 6.118ms  ;  old: 6.830ms   (89.58% )
in: 3x10x32x32; ch: 24; k: 5x7x7; pad: 1; st: 1; new: 58.139ms ;  old: 68.241ms  (85.20% )
in: 3x10x32x32; ch: 24; k: 5x7x7; pad: 1; st: 2; new: 5.413ms  ;  old: 5.598ms   (96.70% )
in: 3x10x32x32; ch: 24; k: 4x4x5; pad: 4; st: 1; new: 60.782ms ;  old: 86.373ms  (70.37% )
in: 3x10x32x32; ch: 24; k: 4x4x5; pad: 4; st: 2; new: 9.972ms  ;  old: 7.348ms   (135.71%)
in: 3x10x32x32; ch: 24; k: 4x4x5; pad: 2; st: 1; new: 36.874ms ;  old: 44.695ms  (82.50% )
in: 3x10x32x32; ch: 24; k: 4x4x5; pad: 2; st: 2; new: 4.657ms  ;  old: 6.031ms   (77.22% )
in: 3x10x32x32; ch: 24; k: 4x4x5; pad: 1; st: 1; new: 18.975ms ;  old: 25.799ms  (73.55% )
in: 3x10x32x32; ch: 24; k: 4x4x5; pad: 1; st: 2; new: 7.087ms  ;  old: 5.713ms   (124.05%)
in: 3x10x32x32; ch: 24; k: 3x3x3; pad: 4; st: 1; new: 27.372ms ;  old: 41.110ms  (66.58% )
in: 3x10x32x32; ch: 24; k: 3x3x3; pad: 4; st: 2; new: 4.503ms  ;  old: 4.507ms   (99.91% )
in: 3x10x32x32; ch: 24; k: 3x3x3; pad: 2; st: 1; new: 16.079ms ;  old: 29.916ms  (53.75% )
in: 3x10x32x32; ch: 24; k: 3x3x3; pad: 2; st: 2; new: 2.727ms  ;  old: 3.250ms   (83.91% )
in: 3x10x32x32; ch: 24; k: 3x3x3; pad: 1; st: 1; new: 14.481ms ;  old: 27.597ms  (52.47% )
in: 3x10x32x32; ch: 24; k: 3x3x3; pad: 1; st: 2; new: 2.174ms  ;  old: 2.839ms   (76.58% )
in: 3x6x16x32;  ch: 32; k: 5x7x7; pad: 4; st: 1; new: 53.375ms ;  old: 67.391ms  (79.20% )
in: 3x6x16x32;  ch: 32; k: 5x7x7; pad: 4; st: 2; new: 5.109ms  ;  old: 6.040ms   (84.59% )
in: 3x6x16x32;  ch: 32; k: 5x7x7; pad: 2; st: 1; new: 18.889ms ;  old: 17.349ms  (108.88%)
in: 3x6x16x32;  ch: 32; k: 5x7x7; pad: 2; st: 2; new: 2.492ms  ;  old: 4.598ms   (54.20% )
in: 3x6x16x32;  ch: 32; k: 5x7x7; pad: 1; st: 1; new: 8.562ms  ;  old: 8.616ms   (99.37% )
in: 3x6x16x32;  ch: 32; k: 5x7x7; pad: 1; st: 2; new: 2.583ms  ;  old: 2.598ms   (99.42% )
in: 3x6x16x32;  ch: 32; k: 4x4x5; pad: 4; st: 1; new: 22.438ms ;  old: 33.031ms  (67.93% )
in: 3x6x16x32;  ch: 32; k: 4x4x5; pad: 4; st: 2; new: 1.836ms  ;  old: 2.666ms   (68.87% )
in: 3x6x16x32;  ch: 32; k: 4x4x5; pad: 2; st: 1; new: 7.689ms  ;  old: 11.028ms  (69.72% )
in: 3x6x16x32;  ch: 32; k: 4x4x5; pad: 2; st: 2; new: 1.705ms  ;  old: 2.567ms   (66.42% )
in: 3x6x16x32;  ch: 32; k: 4x4x5; pad: 1; st: 1; new: 4.510ms  ;  old: 5.739ms   (78.59% )
in: 3x6x16x32;  ch: 32; k: 4x4x5; pad: 1; st: 2; new: 1.262ms  ;  old: 1.451ms   (86.97% )
in: 3x6x16x32;  ch: 32; k: 3x3x3; pad: 4; st: 1; new: 7.719ms  ;  old: 18.303ms  (42.17% )
in: 3x6x16x32;  ch: 32; k: 3x3x3; pad: 4; st: 2; new: 2.110ms  ;  old: 2.684ms   (78.61% )
```

**Comment by zou3519** [comment]
I am unlanding this because it broke the backwards compatibility test on master: https://circleci.com/gh/pytorch/pytorch/3470628?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link/console.

However, I noticed that the test at PR time skips the schema ([full log here](https://circleci.com/gh/pytorch/pytorch/3454673?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link):
```
Nov 04 17:31:09 skipping schema:  aten::thnn_conv3d.out(Tensor self, Tensor weight, int[3] kernel_size, Tensor? bias=None, int[3] stride=[1, 1, 1], int[3] padding=[0, 0, 0], *, Tensor(a!) out) -> (Tensor(a!))
Nov 04 17:31:09 skipping schema:  aten::thnn_conv3d(Tensor self, Tensor weight, int[3] kernel_size, Tensor? bias=None, int[3] stride=[1, 1, 1], int[3] padding=[0, 0, 0]) -> (Tensor)
```
but the schema isn't skipped on the test that is run in master. @houseroad do you know what could be up here?

**Comment by houseroad** [comment]
We need to make sure this PR doesn't break any internal models

**Comment by andreaskoepf** [comment]
Is the primary problem the renaming of the functions? I tried to remove the   thnn legacy. But if that is not easily possible without becoming a BC I could revert the name changes. Please let me know if that is appropriate here!

**Comment by ezyang** [comment]
@andreaskoepf no, it's really dumb and simple, we just need to whitelist the .out name too. I can do this for you.

**Comment by andreaskoepf** [comment]
@ezyang ok, if it is simple and clear to you what to do, please go ahead!

**Comment by ezyang** [comment]
@houseroad So, when I resubmit this patch, how do I make sure master doesn't break, if the PR/master behavior is different?

**Comment by houseroad** [comment]
If we intend to have breaking changes, add them to the whitelist: https://github.com/pytorch/pytorch/blob/master/test/backward_compatibility/check_backward_compatibility.py#L30

**Comment by ezyang** [comment]
@houseroad Err, what I mean is, I thought I had added entries to the whitelist that were sufficient, and the PR build status said that everything was OK, but when I actually landed it, the tests failed on master (see also Richard's comment). Am I holding it wrong or something?

**Comment by houseroad** [comment]
Because the expiration data was set to Nov.4, so the PR looks good, but the master is broken...

**Comment by ezyang** [comment]
@houseroad So... I'm supposed to pick a date that is somewhere far off in the future? That is totally not obvious from the file in question ;)

**Comment by ezyang** [comment]
@houseroad can you please review https://github.com/pytorch/pytorch/pull/29286 it also contains comments in BC test about this case

**Pull request closed without merging**